CNV#45704 4.18: Configuring a storage class for custom boot source updates (#88579)
openshift/openshift-docs: modules/virt-configuring-storage-class-bootsource-update.adoc

:_mod-docs-content-type: PROCEDURE
[id="virt-configuring-storage-class-bootsource-update_{context}"]
= Configuring a storage class for custom boot source updates

You can override the default storage class by editing the `HyperConverged` custom resource (CR).
= Configuring a storage class for boot source images

[IMPORTANT]
====

> Boot sources are created from storage using the default storage class. If your cluster does not have a default storage class, you must define one before configuring automatic updates for custom boot sources.

Reviewer: Is it necessary for this to be in an important block, especially as the introduction?

> 

Reviewer 2: In line 8, it says "custom" boot source updates, but it's relevant for both custom and system-defined boot sources, maybe it can be

`= Configuring a storage class for the boot source images
`

And we can remove 
`You can override the default storage class by editing the HyperConverged custom resource (CR).`
because in line 18 we say that one can specify the `storageClassName`

[IMPORTANT]
====
Boot sources are created using the default storage class. When the default storage class changes, old boot sources are automatically updated using the new default storage class.

If boot source images were stored as VolumeSnapshots and both the cluster default and `virt-default` storage class have been unset, the VolumeSnapshots will be cleaned up and new DataVolumes will be created. However they will not start importing until a default storage class is set.

> 

Reviewer 2: What do you think about adding this new module right here?
`include::modules/virt-configuring-default-and-virt-default-storage-class.adoc[leveloffset=+2]`

So the story flow is:
- Title: Configuring a storage class for the boot source images 
- Boot sources need a default sc, here's how to set it
- You can set the sc in HCO, here's how

> To ensure stable behavior and avoid unnecessary re-importing, you can specify the `storageClassName` in the `dataImportCronTemplates` section of the `HyperConverged` resource.

Reviewer 2: I can't comment on the lines below, in the Procedure section, but we need some changes there as well, because dataImportCronTemplates is not part of the HyperConverged `spec`. But users can add this filed to the spec. 

Also, in the current doc, yaml is wrong - it should be `...spec.storage.storageClassName`

So this would be a better wording with a correct yaml:

Procedure
Open the `HyperConverged` CR in your default editor by running the following command:

$ oc edit hyperconverged kubevirt-hyperconverged -n {CNVNamespace}

Insert the dataImportCronTemplate to the HyperConverged resource's spec, and specify the storageClassName:

```
apiVersion: hco.kubevirt.io/v1beta1
kind: HyperConverged
metadata:
  name: kubevirt-hyperconverged
spec:
  dataImportCronTemplates:
  - metadata:
      name: rhel9-image-cron
    spec:
      template:
        spec:
          storage:
            storageClassName: <storage_class> 
      schedule: "0 */12 * * *" 
      managedDataSource: <data_source> 
```

- Wait For HCO & SSP resources to reconcile

- Delete the related old DataVolume, VolumeSnapshot objects from the openshift-virtualization-os-images namespace.

- Wait for all DataSources to have Ready status.

> To ensure stable behavior and avoid unnecessary re-importing, you can specify the `storageClassName` in the `dataImportCronTemplates` section of the `HyperConverged` resource.

Reviewer 2: Also section about changing the default storage class (starting at line 61) needs a change... It should be logically separated from the editing of the HyperConverged (maybe moved up? Maybe moved to the part of the docs where we explain how to configure the default and virt-default storage classes?). 
And it should mention the virt-default storage class. This would better describe the procedure:

```
Change the cluster default, or a virt-default storage class (virt-default storage class has precedence over the cluster default storage class)
Patch the current virt-default or a cluster-default storage class to “false”

Query all storage classes that are marked as virt-default:

oc get sc -o json| jq '.items[].metadata|select(.annotations."storageclass.kubevirt.io/is-default-virt-class"=="true")|.name'

Patch each of them:

oc patch storageclass <storage_class_name> -p '{"metadata": {"annotations": {"storageclass.kubevirt.io/is-default-virt-class": "false"}}}'

Query all storage classes that are marked as cluster-default:

oc get sc -o json| jq '.items[].metadata|select(.annotations."storageclass.kubernetes.io/is-default-class"=="true")|.name'

Patch each of them:

oc patch storageclass <storage_class_name> -p '{"metadata": {"annotations": {"storageclass.kubernetes.io/is-default-class": "false"}}}'

Patch the desired storage class to be a virt-default or a cluster-default:

oc patch storageclass <storage_class_name> -p '{"metadata": {"annotations": {"storageclass.kubevirt.io/is-default-virt-class": "true"}}}'

or

oc patch storageclass <storage_class_name> -p '{"metadata": {"annotations": {"storageclass.kubernetes.io/is-default-class": "true"}}}'

There should only be one of each.

```

----
--

. Wait for the HyperConverged Operator (HCO) and Scheduling, Scale, and Performance (SSP) resources to complete reconciliation.

> . Delete any outdated DataVolume and VolumeSnapshot objects from the `openshift-virtualization-os-images` namespace.

Reviewer 3: ```suggestion
. Delete any outdated `DataVolume` and `VolumeSnapshot` objects from the `openshift-virtualization-os-images` namespace.
```

> . Delete any outdated `DataVolume` and `VolumeSnapshot` objects from the `openshift-virtualization-os-images` namespace.

Reviewer: ```suggestion
. Delete any outdated `DataVolume` and `VolumeSnapshot` objects from the `openshift-virtualization-os-images` namespace by entering the following command: 
```

+
[source,terminal]
----
$ oc delete DataVolume,VolumeSnapshot -n openshift-virtualization-os-images --selector=cdi.kubevirt.io/dataImportCron
----

> . Wait for all DataSources to reach a "Ready - True" status. DataSources can reference either a PersistentVolumeClaim (PVC) or a VolumeSnapshot. To check the expected source format, run the following command:

Reviewer 3: ```suggestion
. Wait for all `DataSource` objects to reach a "Ready - True" status. Data sources can reference either a PersistentVolumeClaim (PVC) or a VolumeSnapshot. To check the expected source format, run the following command:
```

$ oc patch storageclass <new_storage_class> -p '{"metadata":{"annotations":{"storageclass.kubernetes.io/is-default-class":"true"}}}' <1>
----

> <1> Replace `<new_storage_class>` with the `storageClassName` value that you added to the `HyperConverged` CR.

Reviewer 2: So with the new file about changing the storage class, we can remove from here everything from line 62 to 94 :)
